### PR TITLE
Enable Reprocessing by Default and Add Info Emails

### PR DESCRIPTION
**Pull request by mitsuhiko**
This PR turns on reprocessing on by default for everybody. The old hint is gone and instead it is being replaced by a new call to action email.

Behavior:

* if reprocessing is disabled you get a single email telling you that you can turn it on. This email is only ever sent once after an event ocurred that reprocessing could have fixed. If the email is already sent and you did not turn on reprocessing since you will not get any further emails.
* if reprocessing is enabled you get that email every time a new processing issue happened and you did not resolve it in the meantime. This is needed because you will not receive any alerts otherwise.

### No Reprocessing Active

This mail is only ever sent a single time to prompt the user to turn on reprocessing.

screen shot 2017-04-21 at 09 57 02

The text version:

```
Some events failed to process in your project "My Project".

Issues:

* A required debug symbol file was missing.
  (CrashProbeiOS (arm64))

* The debug symbol file used was broken.
  (libCrashProbeiOS (arm64))


Because reprocessing is not enabled we
will not futher bother you about issues like this.  We do however
recommend to enable reprocessing in the project settings
(http://localhost:8000/organization/project/settings/processing-issues/)
for a better user experience.
```

### Reprocessing Active

This is sent the first time processing issues occur. If they are resolved and happen again you get a new mail

screen shot 2017-04-21 at 09 57 54

Text version:

```
Some events failed to process in your project "My Project".

Because reprocessing is enabled you will not see these events until
you corrected some issues that affect processing.

Issues:

* A required debug symbol file was missing.
  (CrashProbeiOS (arm64))

* The debug symbol file used was broken.
  (libCrashProbeiOS (arm64))


You can fix these issues here:
  http://localhost:8000/organization/project/settings/processing-issues/
```

### CODE REVIEW

**Comment by mitsuhiko** [comment]
@ckj for the contents of that mail it would be great if you could have a look over if that makes sense.

@mattrobenolt the way I implemented the mail targeting is highly questionable. I tried to send the mail to everybody who would normally get an alert for a group in a project however that code assumes that an actual group exists (which it does not do here). So I badly derived the list of users from the UI code which drives the selection checkbox. I would love to clean this up but I'm not sure if I'm missing something there.

File: src/sentry/templates/sentry/emails/activity/new_processing_issues.html [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        </tbody>
+      </table>
+    </div>
+
+    {% if not reprocessing_active %}
+    <p>
+      Because reprocessing is not enabled we will not futher bother you about
+      issues like this.  We do however recommend to <a href="{{ info_url
+        }}">enable reprocessing in the project settings</a> for a better user
+      experience.
```
Comment (ckj) [inline src/sentry/templates/sentry/emails/activity/new_processing_issues.html]: > Because reprocessing is not enabled we will not futher bother you about issues like this.  We do however recommend to enable reprocessing in the project settings for a better user experience.

"Since reprocessing is disabled, we won't send you additional notifications about issues like this. However, we recommend **enabling reprocessing** in project settings to improve your debugging experience on Sentry."

**Review by ckj — CHANGES_REQUESTED** [review]
The language is a little hard to follow. I added some suggestions inline.

**Comment by mitsuhiko** [comment]
@ckj changed those.

**Review by ckj — APPROVED** [review]
👍

File: src/sentry/models/groupsubscription.py [hunk 2]
```diff
@@ -12,8 +12,9 @@
 
 
 class GroupSubscriptionReason(object):
-    committed = -2  # not for use as a persisted field value
     implicit = -1   # not for use as a persisted field value
+    committed = -2  # not for use as a persisted field value
```
Comment (mattrobenolt) [inline src/sentry/models/groupsubscription.py]: I like that you moved these from ascending to descending order. ;)

File: src/sentry/plugins/sentry_mail/activity/new_processing_issues.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+            'extra_info': extra_info,
+        })
+    return rv
+
+
+class NewProcessingIssuesActivityEmail(ActivityEmail):
+
+    def __init__(self, activity):
+        ActivityEmail.__init__(self, activity)
+        self.organization = self.project.organization
```
Comment (mattrobenolt) [inline src/sentry/plugins/sentry_mail/activity/new_processing_issues.py]: Why are you doing this? This is being provided by `ActivityEmail`.

File: src/sentry/plugins/sentry_mail/activity/new_processing_issues.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        # base notify code and requires event information, we use the UI
+        # code (Project.is_user_subscribed_to_mail_alerts) which
+        # replicates the logic on a per-user basis.
+        users = User.objects.filter(
+            sentry_orgmember_set__teams=self.project.team,
+            is_active=True,
+        )
+        participants = {}
+        for user in users:
+            if self.project.is_user_subscribed_to_mail_alerts(user):
```
Comment (mattrobenolt) [inline src/sentry/plugins/sentry_mail/activity/new_processing_issues.py]: This is less than ideal, I agree. Doing this is potentially two queries per user. And there many be hundreds of members on the org. I'd like to see `is_user_subscribed_to_mail_alerts` either be refactored to accept a list of users, or add another method for checking this in bulk. Since in theory, this can be constant time with only 2 queries no matter how many users there are.

File: src/sentry/plugins/sentry_mail/activity/new_processing_issues.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        self.organization = self.project.organization
+        self.issues = summarize_issues(self.data['issues'])
+
+    def get_participants(self):
+        # XXX: We want to send an email to everybody who is subscribed to
+        # the mail alerts.  Since currenlty that is only checked in the
+        # base notify code and requires event information, we use the UI
+        # code (Project.is_user_subscribed_to_mail_alerts) which
+        # replicates the logic on a per-user basis.
+        users = User.objects.filter(
```
Comment (mattrobenolt) [inline src/sentry/plugins/sentry_mail/activity/new_processing_issues.py]: I'd do one of these two options:

```python
users = User.objects.filter(
    id__in=self.project.team.member_set.values_list('user_id'),
)
```

Or:

```python
members = team.member_set.select_related('user')

for member in members:
    user = member.user
    ...
```

Purely to avoid repeating the logic there.

File: src/sentry/tasks/store.py [hunk 6]
```diff
@@ -177,24 +183,66 @@ def delete_raw_event(project_id, event_id):
...
+    reprocessing_active = ProjectOption.objects.get_value(
+        project_id, 'sentry:reprocessing_active', REPROCESSING_DEFAULT)
+    if reprocessing_active:
+        sent_notification = ProjectOption.objects.get_value(
+            project_id, 'sentry:sent_failed_event_hint', False)
+        if sent_notification:
+            if ReprocessingReport.objects.filter(
+                project_id=project_id,
+                event_id=event_id
+            ).values('id').first() is None:
```
Comment (mattrobenolt) [inline src/sentry/tasks/store.py]: ProTip:

`.exists()` translates to a `SELECT 1 ...` and is what you should use here which returns a boolean.

**Review by mattrobenolt — CHANGES_REQUESTED** [review]
Are we going to be shooting ourselves in the foot here by generically naming these things as just "processing issues"? I'm thinking when we do reprocessing for JS, we'd want to split this behavior. Say, send first time per-platform instead of globally, first time ever.

File: src/sentry/plugins/sentry_mail/activity/new_processing_issues.py (hunk 3 above)
Comment (mitsuhiko) [inline src/sentry/plugins/sentry_mail/activity/new_processing_issues.py]: Copy paste from somewhere else. Will investigate.

File: src/sentry/plugins/sentry_mail/activity/new_processing_issues.py (hunk 5 above)
Comment (mitsuhiko) [inline src/sentry/plugins/sentry_mail/activity/new_processing_issues.py]: This is changed now as well.

**Comment by mitsuhiko** [comment]
I am blocking this on this here: https://github.com/getsentry/sentry/issues/5335

**Comment by mitsuhiko** [comment]
The user targeting should be more efficient now and more reusable.

File: src/sentry/models/project.py [hunk 7]
```diff
@@ -260,6 +260,43 @@ def get_full_name(self):
...
+
+        member_set = set(self.member_set.exclude(
+            user__in=disabled,
+        ).values_list('user', flat=True))
+
+        # determine members default settings
+        members_to_check = set(u for u in member_set if u not in alert_settings)
+        if members_to_check:
+            disabled = set((
+                uo.user_id for uo in UserOption.objects.filter(
```
Comment (mattrobenolt) [inline src/sentry/models/project.py]: This query needs to be qualified with a project. Otherwise, this doesn't hit an index.

File: CHANGES [hunk 8]
```diff
@@ -8,6 +8,7 @@ Version 8.16
 
 - Added data migration to create UserEmail objects for users whose primary emails did not have them
 - Time series data (used by graphs and other features) is now updated when groups are merged.
+- Turned on reprocessing by default
```
Comment (mattrobenolt) [inline CHANGES]: Under wrong version.

File: src/sentry/models/project.py (hunk 7 above)
Comment (mattrobenolt) [inline src/sentry/models/project.py]: jk I'm wrong. This is fine.

**Review by mattrobenolt — APPROVED** [review]
